src/filerepository.js

```diff
@@ -0,0 +1,455 @@
+/**
+ * @license Copyright (c) 2003-2017, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module upload/filerepository
+ */
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+
+import CKEditorError from '@ckeditor/ckeditor5-utils/src/ckeditorerror';
+import ObservableMixin from '@ckeditor/ckeditor5-utils/src/observablemixin';
+import Collection from '@ckeditor/ckeditor5-utils/src/collection';
+import mix from '@ckeditor/ckeditor5-utils/src/mix';
+import log from '@ckeditor/ckeditor5-utils/src/log';
+
+import FileReader from './filereader.js';
+
+import uid from '@ckeditor/ckeditor5-utils/src/uid.js';
+
+/**
+ * FileRepository plugin.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class FileRepository extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'fileRepository';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		/**
+		 * Collection of loaders associated with this repository.
+		 *
+		 * @member {module:utils/collection~Collection} #loaders
+		 */
+		this.loaders = new Collection();
+
+		/**
+		 * Function that should be defined before using FileRepository. It should return new instance of
+		 * {@link module:upload/filerepository~Adapter Adapter} that will be used to upload files.
+		 * {@link module:upload/filerepository~FileLoader FileLoader} instance associated with the adapter
+		 * will be passed to that function.
+		 *
+		 *	fileRepository.createAdapter = function( loader ) {
+		 *		return new ServerAdapter();
```

Comment:
It's not very useful for someone who doesn't know how `ServerAdapter` should look like and someone who knows how `ServerAdapter` should look like should also know about this. I think it will be more useful for the developer to keep both together.